spacy_llm/pipeline/llm.py

```diff
@@ -232,15 +237,22 @@ def pipe(self, stream: Iterable[Doc], *, batch_size: int = 128) -> Iterator[Doc]
         batch_size (int): The number of documents to buffer.
         YIELDS (Doc): Processed documents in order.
         """
+        if not Doc.has_extension("llm_io"):
+            Doc.set_extension("llm_io", default=None)
+
         error_handler = self.get_error_handler()
         for doc_batch in spacy.util.minibatch(stream, batch_size):
             is_cached = [doc in self._cache for doc in doc_batch]
             noncached_doc_batch = [
                 doc for doc, cached_doc in zip(doc_batch, is_cached) if not cached_doc
             ]
             try:
-                prompts = self._task.generate_prompts(noncached_doc_batch)
-                responses = self._backend(prompts)
+                prompts = list(self._task.generate_prompts(noncached_doc_batch))
+                iter_prompts = iter(prompts)
+
+                responses = list(self._backend(prompts))
+                iter_responses = iter(responses)
```

Comment:
Why the list conversion?

Reply:
Otherwise the iterator is consumed by `self._backend(prompts)`. Unless using `iter` makes it unnecessary?

Reply:
I think what you're looking for might be `itertools.tee`?

Reply:
Ah, you're using the prompts and responses for the logging again. Give it a shot with `itertools.tee()`?

Reply:
Since they are consumed sequentially [I'm not sure it will help](https://docs.python.org/3/library/itertools.html#itertools.tee:~:text=This%20itertool%20may%20require%20significant%20auxiliary%20storage%20(depending%20on%20how%20much%20temporary%20data%20needs%20to%20be%20stored).%20In%20general%2C%20if%20one%20iterator%20uses%20most%20or%20all%20of%20the%20data%20before%20another%20iterator%20starts%2C%20it%20is%20faster%20to%20use%20list()%20instead%20of%20tee().)...

Reply:
It's not that this doesn't work, but I think it's more confusing at first glance than using `itertools.tee`.

I do worry in general that this whole collection of multiple interleaved steps with iterators is very brittle and likely to get out-of-sync.

Reply:
It might also be clearer if the second iterators are only created if required?

Reply:
I agree. We could also defer the responsibility to the task... I did it this way because it does not require any change on the task, but come to think of it it would make sense that they handle it.

Consider the case where the `Prompt` is a complex, non-serialisable object... It should be handled differently.

Reply:
Ah, and I also struggled to include it at the task level _and_ get the component name...

Reply:
> Consider the case where the Prompt is a complex, non-serialisable object... It should be handled differently.

We could only store the stringified version of prompts and responses?

Reply:
> Ah, and I also struggled to include it at the task level and get the component name...

Can you elaborate on this?

Reply:
> We could only store the stringified version of prompts and responses?

Good idea. Then it's the task's/backends responsibility to generate prompts/responses that display a proper `__str__` method.

Reply:
I looked at a potential solution to implement prompt/response storage at the task level, but it looks very convoluted. We'd likely need to apply substantial changes to the architecture.

> Ah, and I also struggled to include it at the task level and get the component name...

Storage is done on a per-component basis so we need the component name. We could use a list or more exotic workarounds, but I think that's more confusing than anything else. Do you have another idea?

Reply:
Ah, ok. I think it's fine to do this at the component level. I don't think customized logging (via tasks) is a priority right now.